apollo-runtime/src/appleMain/kotlin/com/apollographql/apollo3/network/http/StreamingNSURLSessionHttpEngine.kt

```diff
@@ -0,0 +1,313 @@
+    fun onComplete(error: NSError?)
+  }
+
+  private val handlers = mutableMapOf<NSURLSessionTask, Handler>()
+
+  override fun URLSession(
+      session: NSURLSession,
+      dataTask: NSURLSessionDataTask,
+      didReceiveResponse: NSURLResponse,
+      completionHandler: (NSURLSessionResponseDisposition) -> Unit,
+  ) {
+    handlers[dataTask]?.onResponse(didReceiveResponse as NSHTTPURLResponse)
+    completionHandler(NSURLSessionResponseAllow)
+  }
+
+  override fun URLSession(
+      session: NSURLSession,
+      dataTask: NSURLSessionDataTask,
+      didReceiveData: NSData,
+  ) {
+    handlers[dataTask]?.onData(didReceiveData)
+  }
+
+  override fun URLSession(
+      session: NSURLSession,
+      task: NSURLSessionTask,
+      didCompleteWithError: NSError?,
+  ) {
+    handlers[task]?.onComplete(didCompleteWithError)
+
+    // Cleanup
+    handlers.remove(task)
+  }
+
+  override fun URLSession(
+      session: NSURLSession,
+      dataTask: NSURLSessionDataTask,
+      willCacheResponse: NSCachedURLResponse,
+      completionHandler: (NSCachedURLResponse?) -> Unit,
+  ) {
+    handlers[dataTask]?.onComplete(null)
+
+    // Cleanup
+    handlers.remove(dataTask)
+  }
+
+  fun registerHandlerForTask(task: NSURLSessionTask, handler: Handler) {
+    handlers[task] = handler
+  }
+}
+
+/**
+ * A source and a sink that are attached. The sink's output is the source's input.
+ * A [Buffer] is used internally to store the data.
+ * A `pthread` `mutex` and `cond` are used to block the source's reading until there is data to read, or the sink is closed. These are
+ * released when the **source** is closed.
+ *
+ * Inspired by okio's Pipe which is JVM only.
+ */
+private class Pipe {
+  private val buffer = Buffer()
+
+  private val mutex = nativeHeap.alloc<pthread_mutex_t>()
+  private val cond = nativeHeap.alloc<pthread_cond_t>()
+
+  private var isSinkClosed = false
+
+  init {
+    pthread_mutex_init(mutex.ptr, null)
+    pthread_cond_init(cond.ptr, null)
+  }
+
+  val sink: Sink = object : Sink {
+    override fun write(source: Buffer, byteCount: Long) {
+      pthread_mutex_lock(mutex.ptr)
+      buffer.write(source, byteCount)
+      pthread_cond_broadcast(cond.ptr)
+      pthread_mutex_unlock(mutex.ptr)
+    }
+
+    override fun close() {
+      pthread_mutex_lock(mutex.ptr)
+      isSinkClosed = true
+      pthread_cond_broadcast(cond.ptr)
+      pthread_mutex_unlock(mutex.ptr)
+    }
+
+    override fun flush() {}
+    override fun timeout() = Timeout.NONE
+  }
+
+  val source: Source = object : Source {
+    override fun read(sink: Buffer, byteCount: Long): Long {
+      pthread_mutex_lock(mutex.ptr)
+      while (buffer.size == 0L && !isSinkClosed) {
+        pthread_cond_wait(cond.ptr, mutex.ptr)
+      }
+      val readCount = buffer.read(sink, byteCount)
+      pthread_mutex_unlock(mutex.ptr)
+      return readCount
+    }
+
+    override fun close() {
+      pthread_mutex_destroy(mutex.ptr)
+      pthread_cond_destroy(cond.ptr)
```

Comment:
This is potentially dangerous if the source is closed while the request is still ongoing. I don't really see a way around waiting for both the sink & the source to be closed before releasing the mutex & cond. That's more hoops but I don't think we can make the assumption that the source is going to be closed last?